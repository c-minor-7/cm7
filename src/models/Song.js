import Section from './Section';
import createEl from '../helpers/createEl';

import { findFirstChildOfType } from '../helpers/AST';

export default class Song {
  constructor({ configs, sections }) {
    Object.assign(this, { configs, sections });
  }

  toDOM({ cssClasses }) {
    const { key } = this.configs;
    return createEl(`div.${cssClasses.song}`, {
      children: this.sections.map(
        section => section.toDOM({ key, cssClasses })
      ),
    });
  }

  static fromAST(ast) {
    const configs = findFirstChildOfType(ast, 'configs')
      .text.split('\n')
      .map(config => config.split('='))
      .reduce((acc, [key, value]) => ({
        ...acc,
        [key.trim()]: value.trim(),
      }), {});

    const sections = findFirstChildOfType(ast, 'song').children.map(Section.fromAST);

    return new Song({ configs, sections });
  }
}
