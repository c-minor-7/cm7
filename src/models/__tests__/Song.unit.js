import { readFileSync, readdirSync } from 'fs';
import { join } from 'path';
import pretty from 'pretty';

import parseCm7 from 'cm7-lang-parser';
import Song from '../Song';
import cssClasses from '../../defaultCssClasses';

describe('cm7', () => {
  const cm7sDir = join(__dirname, './cm7s');

  for (const filename of readdirSync(cm7sDir)) {
    if (filename[0] === '.') continue;

    describe(filename, () => {
      it('should generate html that matches snapshot', async () => {
        const src = readFileSync(join(cm7sDir, 'here-we-are.cm7'), 'utf8');

        const songDOM = Song.fromAST(parseCm7(src)).toDOM({ cssClasses });
        const innerHTML = songDOM.innerHTML;
        expect(pretty(innerHTML)).toMatchSnapshot();
      });
    });
  }
});
