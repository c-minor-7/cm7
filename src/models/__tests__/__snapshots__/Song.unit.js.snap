// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cm7 here-we-are.cm7 should generate html that matches snapshot 1`] = `
"<div class=\\"cm7_section\\">
  <div class=\\"cm7_section_label\\">intro</div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">D</span><span class=\\"cm7_chord\\">G</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">D</span><span class=\\"cm7_chord\\">G</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_lyrics\\">:: verse ::</div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">D</span><span class=\\"cm7_chord\\">G</span></div>
    <div class=\\"cm7_line_lyrics\\">()<span class=\\"cm7_line_lyrics-beat\\"> Here we are in the </span>(future)<span class=\\"cm7_line_lyrics-beat\\"> now</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">D</span><span class=\\"cm7_chord\\">G</span><span class=\\"cm7_chord\\">Em</span></div>
    <div class=\\"cm7_line_lyrics\\">()<span class=\\"cm7_line_lyrics-beat\\"> Here we are in the </span>(place)<span class=\\"cm7_line_lyrics-beat\\"> that we wanted to </span>(be)</div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">C</span></div>
    <div class=\\"cm7_line_lyrics\\">To <span class=\\"cm7_line_lyrics-beat\\">(be)</span></div>
  </div>
</div>
<div class=\\"cm7_section\\">
  <div class=\\"cm7_section_label\\"></div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">D</span><span class=\\"cm7_chord\\">G</span></div>
    <div class=\\"cm7_line_lyrics\\">()<span class=\\"cm7_line_lyrics-beat\\"> Everything in your </span>(fantasy)</div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">D</span><span class=\\"cm7_chord\\">G</span><span class=\\"cm7_chord\\">Em</span></div>
    <div class=\\"cm7_line_lyrics\\">()<span class=\\"cm7_line_lyrics-beat\\"> Everything has be</span>(come)<span class=\\"cm7_line_lyrics-beat\\"> your reali</span>(ty)</div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">C</span></div>
    <div class=\\"cm7_line_lyrics\\">(Hmmm)</div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_lyrics\\">:: pre-chorus ::</div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">G</span><span class=\\"cm7_chord\\">A</span></div>
    <div class=\\"cm7_line_lyrics\\">()<span class=\\"cm7_line_lyrics-beat\\"> What a fool I </span>(used)<span class=\\"cm7_line_lyrics-beat\\"> to be</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">G</span><span class=\\"cm7_chord\\">A</span></div>
    <div class=\\"cm7_line_lyrics\\">()<span class=\\"cm7_line_lyrics-beat\\"> Should have known you mean the </span>(most)<span class=\\"cm7_line_lyrics-beat\\"> to me</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">G</span><span class=\\"cm7_chord\\">G</span></div>
    <div class=\\"cm7_line_lyrics\\">()<span class=\\"cm7_line_lyrics-beat\\"> Now I finally </span>(see)<span class=\\"cm7_line_lyrics-beat\\"> it</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">Em</span></div>
    <div class=\\"cm7_line_lyrics\\">Now I finally <span class=\\"cm7_line_lyrics-beat\\">(know)</span> it</div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_lyrics\\">All I really want</div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">Em</span></div>
    <div class=\\"cm7_line_lyrics\\">()<span class=\\"cm7_line_lyrics-beat\\"> All I really want is</span></div>
  </div>
</div>
<div class=\\"cm7_section\\">
  <div class=\\"cm7_section_label\\">chorus</div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">D</span></div>
    <div class=\\"cm7_line_lyrics\\">(I)<span class=\\"cm7_line_lyrics-beat\\"> just wanna be with you</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">D</span></div>
    <div class=\\"cm7_line_lyrics\\">(I)<span class=\\"cm7_line_lyrics-beat\\"> just wanna be with you</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">G</span></div>
    <div class=\\"cm7_line_lyrics\\">Right <span class=\\"cm7_line_lyrics-beat\\">(now)</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">D</span></div>
    <div class=\\"cm7_line_lyrics\\">(I)<span class=\\"cm7_line_lyrics-beat\\"> just wanna be with you</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">D</span></div>
    <div class=\\"cm7_line_lyrics\\">(I)<span class=\\"cm7_line_lyrics-beat\\"> just wanna be with you</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">G</span></div>
    <div class=\\"cm7_line_lyrics\\">Right <span class=\\"cm7_line_lyrics-beat\\">(now)</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">E</span></div>
    <div class=\\"cm7_line_lyrics\\">I <span class=\\"cm7_line_lyrics-beat\\">(didn't)</span> know</div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">Em</span></div>
    <div class=\\"cm7_line_lyrics\\">It <span class=\\"cm7_line_lyrics-beat\\">(runs)</span> so fast</div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">Em7b5</span></div>
    <div class=\\"cm7_line_lyrics\\">And I <span class=\\"cm7_line_lyrics-beat\\">(can't)</span> catch up</div>
  </div>
</div>"
`;

exports[`cm7 london-bridge-is-falling-down.cm7 should generate html that matches snapshot 1`] = `
"<div class=\\"cm7_section\\">
  <div class=\\"cm7_section_label\\">intro</div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">D</span><span class=\\"cm7_chord\\">G</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">D</span><span class=\\"cm7_chord\\">G</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_lyrics\\">:: verse ::</div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">D</span><span class=\\"cm7_chord\\">G</span></div>
    <div class=\\"cm7_line_lyrics\\">()<span class=\\"cm7_line_lyrics-beat\\"> Here we are in the </span>(future)<span class=\\"cm7_line_lyrics-beat\\"> now</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">D</span><span class=\\"cm7_chord\\">G</span><span class=\\"cm7_chord\\">Em</span></div>
    <div class=\\"cm7_line_lyrics\\">()<span class=\\"cm7_line_lyrics-beat\\"> Here we are in the </span>(place)<span class=\\"cm7_line_lyrics-beat\\"> that we wanted to </span>(be)</div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">C</span></div>
    <div class=\\"cm7_line_lyrics\\">To <span class=\\"cm7_line_lyrics-beat\\">(be)</span></div>
  </div>
</div>
<div class=\\"cm7_section\\">
  <div class=\\"cm7_section_label\\"></div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">D</span><span class=\\"cm7_chord\\">G</span></div>
    <div class=\\"cm7_line_lyrics\\">()<span class=\\"cm7_line_lyrics-beat\\"> Everything in your </span>(fantasy)</div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">D</span><span class=\\"cm7_chord\\">G</span><span class=\\"cm7_chord\\">Em</span></div>
    <div class=\\"cm7_line_lyrics\\">()<span class=\\"cm7_line_lyrics-beat\\"> Everything has be</span>(come)<span class=\\"cm7_line_lyrics-beat\\"> your reali</span>(ty)</div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">C</span></div>
    <div class=\\"cm7_line_lyrics\\">(Hmmm)</div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_lyrics\\">:: pre-chorus ::</div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">G</span><span class=\\"cm7_chord\\">A</span></div>
    <div class=\\"cm7_line_lyrics\\">()<span class=\\"cm7_line_lyrics-beat\\"> What a fool I </span>(used)<span class=\\"cm7_line_lyrics-beat\\"> to be</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">G</span><span class=\\"cm7_chord\\">A</span></div>
    <div class=\\"cm7_line_lyrics\\">()<span class=\\"cm7_line_lyrics-beat\\"> Should have known you mean the </span>(most)<span class=\\"cm7_line_lyrics-beat\\"> to me</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">G</span><span class=\\"cm7_chord\\">G</span></div>
    <div class=\\"cm7_line_lyrics\\">()<span class=\\"cm7_line_lyrics-beat\\"> Now I finally </span>(see)<span class=\\"cm7_line_lyrics-beat\\"> it</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">Em</span></div>
    <div class=\\"cm7_line_lyrics\\">Now I finally <span class=\\"cm7_line_lyrics-beat\\">(know)</span> it</div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_lyrics\\">All I really want</div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">Em</span></div>
    <div class=\\"cm7_line_lyrics\\">()<span class=\\"cm7_line_lyrics-beat\\"> All I really want is</span></div>
  </div>
</div>
<div class=\\"cm7_section\\">
  <div class=\\"cm7_section_label\\">chorus</div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">D</span></div>
    <div class=\\"cm7_line_lyrics\\">(I)<span class=\\"cm7_line_lyrics-beat\\"> just wanna be with you</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">D</span></div>
    <div class=\\"cm7_line_lyrics\\">(I)<span class=\\"cm7_line_lyrics-beat\\"> just wanna be with you</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">G</span></div>
    <div class=\\"cm7_line_lyrics\\">Right <span class=\\"cm7_line_lyrics-beat\\">(now)</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">D</span></div>
    <div class=\\"cm7_line_lyrics\\">(I)<span class=\\"cm7_line_lyrics-beat\\"> just wanna be with you</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">D</span></div>
    <div class=\\"cm7_line_lyrics\\">(I)<span class=\\"cm7_line_lyrics-beat\\"> just wanna be with you</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">G</span></div>
    <div class=\\"cm7_line_lyrics\\">Right <span class=\\"cm7_line_lyrics-beat\\">(now)</span></div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">E</span></div>
    <div class=\\"cm7_line_lyrics\\">I <span class=\\"cm7_line_lyrics-beat\\">(didn't)</span> know</div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">Em</span></div>
    <div class=\\"cm7_line_lyrics\\">It <span class=\\"cm7_line_lyrics-beat\\">(runs)</span> so fast</div>
  </div>
  <div class=\\"cm7_line\\">
    <div class=\\"cm7_line_chords\\"><span class=\\"cm7_chord\\">Em7b5</span></div>
    <div class=\\"cm7_line_lyrics\\">And I <span class=\\"cm7_line_lyrics-beat\\">(can't)</span> catch up</div>
  </div>
</div>"
`;
